### PR TITLE
WIP - Stream results from the database to JSON

### PR DESCRIPTION
**Pull request by senior**
This commit uses reducible query results from clojure.java.jdbc to
stream application database query results from to the response as
JSON. Previously we streamed the JSON serialization part, but not the
query portion.

This commit is more intended as an RFC and only switches over the
`/database/:id/fields` endpoint to use this. Testing on my machine, on
large synthetic schema, shows it uses about half the memory that the
previous code used and executes in roughly the same wall clock time.

This commit relies on related query-reducible functionality in Toucan
that hasn't been merged yet, hence the dependency on Toucan
1.1.4-SNAPSHOT.

### CODE REVIEW

File: src/metabase/api/database.clj [hunk 1]
```diff
@@ -261,22 +270,116 @@
...
+  (comp (filter mi/can-read?)
+        ;; This will create 10,000 row chunks of data, being eager for that 10,000 row chunk. This will allow the
+        ;; hydrate calls to be streamed along with the regular database results. Not sure what the best value is for
+        ;; this. A lower value is slower but more memory efficient, a higher value is faster but will consume more
+        ;; memory
+        (partition-all 10000)
+        ;; Hydrate in 10,000 row batches
+        (map #(hydrate % :table))
+        ;; This will flatten the chunks about out
+        cat
```
Comment (sbelak) [inline src/metabase/api/database.clj]: Any reason why not mapcat?

File: src/metabase/api/database.clj [hunk 2]
```diff
@@ -261,22 +270,116 @@
...
+                 ^JsonGenerator json-generator (json/create-generator buffered-writer)]
+       (try
+         ;; We're always going to return a seq of objects, writing them in this way is the fastest way I've found
+         (.writeStartArray json-generator)
+         ;; This is a side-affecty way to run a reducible
+         (run! #(json-gen/encode-map % json-generator)
+               ;; This eduction will help us know that we've successfully executed a query and recevied a result from
+               ;; the result set. Up to this point, we have established a connection to the database but not run the
+               ;; query. We'll check for the first row first before running the `transducer` on the results
+               (eduction (map (call-on-first-row on-first-row)) transducer reducible-q))
```
Comment (sbelak) [inline src/metabase/api/database.clj]: Why do you need eduction here? If I follow, the reducible is immediately (and only once) consumed by run! Can't you simply do:
```
(transduce (comp (map (call-on-first-row on-first-row)) transducer) reducible-q)
```

Comment (senior) [reply src/metabase/api/database.clj]: Seems like an equal trade there. I guess the benefits of the above approach is that it doesn't wrap the reducible in the defrecord. I'd be curious what you think about this being nested in a with-open vs. the side-affecty transducer handling the opening/closing of the writer.

File: src/metabase/api/database.clj (hunk 1 above)
Comment (senior) [reply src/metabase/api/database.clj]: `mapcat` is better here 👍

File: src/metabase/api/database.clj (hunk 2 above)
Comment (sbelak) [reply src/metabase/api/database.clj]: If all things were equal, I'd stuff all state in the transducer, however your current approach is more robust in how it handles exceptions. If all state handling is in the transducer and there's an exception in the xf, you're left dangling.

File: src/metabase/api/database.clj [hunk 3]
```diff
@@ -261,22 +270,116 @@
...
+  (let [p        (promise)
+        response (serialize-to-piped-input-stream #(deliver p %) (make-error-handler p) fields-xform
+                                                  (db/select-reducible [Field :id :display_name :table_id :base_type :special_type]
+                                                    :table_id        [:in (db/select-field :id Table, :db_id id)]
+                                                    :visibility_type [:not-in ["sensitive" "retired"]]))]
+
+    ;; Block until a row or an exception has been delivered to `p`. This will catch something like SQL query that
+    ;; fails to compile or references something that doesn't exist and allow that exception to be returned as if the
+    ;; query was running in the HTTP thread.
+    (throw-if-exception p)
```
Comment (sbelak) [inline src/metabase/api/database.clj]: This is going to be super subjective: but this approach with promises and cross-thread communication is making my brain hurt. Couldn't this be done with something state-machine like?

**Comment by senior** [comment]
Added a new `defendpoint-streaming` that packages up the previous code into a better api. I think `defendpoint` could use a refactor too, but this doesn't feel like the right time to do that.

**Comment by camsaul** [comment]
Closing since @senior is not longer actively working on Metabase 😢 💔

**Pull request closed without merging**